hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/predicate/NotTest.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.ConstantExpression;
+import com.hazelcast.sql.impl.expression.ExpressionTestBase;
+import com.hazelcast.sql.impl.expression.SimpleExpressionEvalContext;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.apache.calcite.rel.type.RelDataType;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.sql.impl.calcite.validate.HazelcastSqlOperatorTable.NOT;
+import static org.apache.calcite.sql.type.SqlTypeName.BOOLEAN;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class NotTest extends ExpressionTestBase {
```

Comment:
Incorrect error messages is produced for some tests:
```
id: 69
query:  SELECT NOT 9223372036854775808.01 FROM t
From line 1, column 8 to line 1, column 33: Cannot apply 'NOT' to arguments of type 'NOT<DOUBLE>'. Supported form(s): 'NOT<BOOLEAN>'
```

Reply:
What's wrong with it?

Reply:
If you talking about `type 'NOT<DOUBLE>'` part, looks like that is some bug in Calcite. I get used to it to the point it looks pretty normal to me now :)

Reply:
Yes, I mean this `NOT` before the type. It is part of the Calcite, but maybe we can hook there and fix that. Otherwise, we confuse our users.

Reply:
Looks like there is no way to customize the message without altering the Calcite code, see `SqlCallBinding.newValidationSignatureError`.

Reply:
It kind of look OKish for expression with more than one operand: `Cannot apply 'AND' to arguments of type '<BIGINT> AND <DOUBLE>'. Supported form(s): '<BOOLEAN> AND <BOOLEAN>'`, but I agree for one-operand expressions it looks especially confusing.

Reply:
AFAIU `SqlCallBinding.newValidationSignatureError` delegates to a specific operator to get this information. Can we override NOT operator to fix the error message?

Reply:
I don't think that fixing it for a single operator makes any sense, it's the general format used by Calcite for all operators. It substitutes a type name surrounded by angle brackets for every formal operand of an operator.

Reply:
What is your proposal wrt this issue?

Reply:
Is it an issue after all? If it is, how critical is it? What is your idea about a better message?

The only potentially universal solution I have in mind for now is overriding `SqlValidatorImpl.newValidationError` and trying to intercept incoming `Resources.ExInst`. Looks like that would require some significant amount of hackery, like classes declared on Calcite packages to gain access to protected fields.

Reply:
From the user perspective, it is confusing - the system reports types that do not exist. At first glance, it doesn't seem difficult to fix. I was able to intercept the signature error by adding the following wrapper to the `HazelcastOperandTypes.NotAny#checkOperandTypes`:
```
SqlCallBinding binding1 = new SqlCallBinding(binding.getValidator(), binding.getScope(),binding.getCall()) {
    @Override
    public CalciteException newValidationSignatureError() {
        return super.newValidationSignatureError();
    }
};
```
Once the call is intercepted, we may produce any error message we like.

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17301 for 4.2